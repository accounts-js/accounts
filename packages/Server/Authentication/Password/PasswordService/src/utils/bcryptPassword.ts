import * as bcrypt from 'bcryptjs';


export const bcryptPassword = async ( password: string ): Promise <string> => {

  const salt = await bcrypt.genSalt(10);

  const hash = await bcrypt.hash(password, salt);

  return hash;

};