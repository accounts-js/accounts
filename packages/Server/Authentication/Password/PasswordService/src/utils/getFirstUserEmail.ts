import { EmailRecord, User } from 'accounts';


export const getFirstUserEmail = ( user: User, address: string ): string => {

  // Pick the first email if we weren't passed an email
  if (!address && user.emails && user.emails[0]) {
    address = user.emails[0].address;
  }

  if(!address) throw new Error('No such email address for user');

  // Make sure the address is valid
  const userEmailRecords: EmailRecord[] = user.emails || [];

  const userEmails = userEmailRecords.map( (email: EmailRecord) => email.address );

  const doAddressBelongsToUser = userEmails.includes(address);

  if(!doAddressBelongsToUser) throw new Error('The address does not belongs to the user');

  return address;

}
