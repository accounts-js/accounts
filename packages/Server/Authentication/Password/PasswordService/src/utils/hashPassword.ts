import { Password } from '../types/Password';

import * as crypto from 'crypto';


export const getHashPassword = ( algorithm: string ) => ( password: Password ) => {

    if (typeof password !== 'string') return password.digest;

    const hash = crypto.createHash(algorithm);

    hash.update(password);

    return hash.digest('hex');
    
}
