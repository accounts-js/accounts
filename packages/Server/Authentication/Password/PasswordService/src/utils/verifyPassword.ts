import * as bcrypt from 'bcryptjs';


export const verifyPassword = async ( password: string, hash: string ): Promise <boolean> => bcrypt.compare(password, hash);
