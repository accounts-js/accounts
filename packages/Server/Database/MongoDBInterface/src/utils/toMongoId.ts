import { ObjectID } from 'mongodb';

export const toMongoID = ( objectId ) => 
typeof objectId === 'string' 
? new ObjectID(objectId) 
: objectId