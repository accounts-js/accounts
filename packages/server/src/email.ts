import { UserObjectType } from '@accounts/common';

export interface EmailTemplateType {
  from?: string;
  subject: (user?: UserObjectType) => string;
  text: (user: UserObjectType, url: string) => string;
}

export interface EmailTemplatesType {
  from: string;
  verifyEmail: EmailTemplateType;
  resetPassword: EmailTemplateType;
  enrollAccount: EmailTemplateType;
}

export const emailTemplates = {
  from: 'js-accounts <no-reply@js-accounts.com>',

  verifyEmail: {
    subject: () => 'Verify your account email',
    text: (user: UserObjectType, url: string) =>
      `To verify your account email please click on this link: ${url}`,
  },

  resetPassword: {
    subject: () => 'Reset your password',
    text: (user: UserObjectType, url: string) =>
      `To reset your password please click on this link: ${url}`,
  },

  enrollAccount: {
    subject: () => 'Set your password',
    text: (user: UserObjectType, url: string) =>
      `To set your password please click on this link: ${url}`,
  },
};

export type SendMailType = (mail: object) => Promise<void>;

export const sendMail = async (mail: object): Promise<void> => {
  // tslint:disable-next-line no-console
  console.warn('No configuration for email, you must set an email configuration');
  // tslint:disable-next-line no-console
  console.log(mail);
};
