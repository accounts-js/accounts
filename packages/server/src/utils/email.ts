import { User } from '@accounts/types';
import { EmailTemplatesType } from '../types/email-templates-type';

export const emailTemplates: EmailTemplatesType = {
  from: 'accounts-js <no-reply@accounts-js.com>',

  verifyEmail: {
    subject: () => 'Verify your account email',
    text: (user: User, url: string) =>
      `To verify your account email please click on this link: ${url}`,
    html: (user: User, url: string) =>
      `To verify your account email please <a href="${url}">click here</a>.`,
  },

  resetPassword: {
    subject: () => 'Reset your password',
    text: (user: User, url: string) => `To reset your password please click on this link: ${url}`,
    html: (user: User, url: string) =>
      `To reset your password please <a href="${url}">click here</a>.`,
  },

  enrollAccount: {
    subject: () => 'Set your password',
    text: (user: User, url: string) => `To set your password please click on this link: ${url}`,
    html: (user: User, url: string) =>
      `To set your password please <a href="${url}">click here</a>.`,
  },
};

export type SendMailType = (mail: object) => Promise<void>;

export const sendMail = async (mail: object): Promise<void> => {
  // tslint:disable-next-line no-console
  console.warn('No configuration for email, you must set an email configuration');
  // tslint:disable-next-line no-console
  console.log(mail);
};
