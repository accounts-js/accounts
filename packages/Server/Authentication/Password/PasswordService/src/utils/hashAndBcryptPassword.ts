import { HashAlgorithm } from "../types/HashAlgorithm";
import { Password } from "../types/Password";

import { bcryptPassword } from "./bcryptPassword";


export const getHashAndBcryptPassword = ( hashPasswordWithAlgorithm: Function ) => ( password: Password ) : Promise <string> => {

    const hashedPassword: string = hashPasswordWithAlgorithm( password );

    return bcryptPassword( hashedPassword )

}