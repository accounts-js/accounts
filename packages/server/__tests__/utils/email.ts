import { emailTemplates } from '../../src/utils/email';

describe('email', () => {
  describe('emailTemplates', () => {
    it('should return default subject', () => {
      expect(emailTemplates.verifyEmail.subject()).toMatchSnapshot();
      expect(emailTemplates.resetPassword.subject()).toMatchSnapshot();
      expect(emailTemplates.enrollAccount.subject()).toMatchSnapshot();
    });

    it('should return default text', () => {
      const user: any = {};
      const url = 'url';
      expect(emailTemplates.verifyEmail.text(user, url)).toMatchSnapshot();
      expect(emailTemplates.resetPassword.text(user, url)).toMatchSnapshot();
      expect(emailTemplates.enrollAccount.text(user, url)).toMatchSnapshot();
    });

    it('should return default html', () => {
      const user: any = {};
      const url = 'url';
      expect(emailTemplates.verifyEmail.html(user, url)).toMatchSnapshot();
      expect(emailTemplates.resetPassword.html(user, url)).toMatchSnapshot();
      expect(emailTemplates.enrollAccount.html(user, url)).toMatchSnapshot();
    });
  });
});
