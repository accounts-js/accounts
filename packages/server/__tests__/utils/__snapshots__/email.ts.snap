// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email emailTemplates should return default html 1`] = `"To verify your account email please <a href=\\"url\\">click here</a>."`;

exports[`email emailTemplates should return default html 2`] = `"To reset your password please <a href=\\"url\\">click here</a>."`;

exports[`email emailTemplates should return default html 3`] = `"To set your password please <a href=\\"url\\">click here</a>."`;

exports[`email emailTemplates should return default subject 1`] = `"Verify your account email"`;

exports[`email emailTemplates should return default subject 2`] = `"Reset your password"`;

exports[`email emailTemplates should return default subject 3`] = `"Set your password"`;

exports[`email emailTemplates should return default text 1`] = `"To verify your account email please click on this link: url"`;

exports[`email emailTemplates should return default text 2`] = `"To reset your password please click on this link: url"`;

exports[`email emailTemplates should return default text 3`] = `"To set your password please click on this link: url"`;
